src/TimeTrace/ExecutionHierarchy.cpp

```diff
@@ -0,0 +1,342 @@
+    }
+}
+
+void ExecutionHierarchy::OnSymbolName(const SymbolName& symbolName)
+{
+    // SymbolName events get executed after all TemplateInstantiation in the same FrontEndPass take place
+    // and we're sure to have exclusive keys for these symbols (they may have matching names between
+    // FrontEndPass activities, but their key is unique for the trace)
+    assert(symbolNames_.find(symbolName.Key()) == symbolNames_.end());
+    std::string& name = symbolNames_[symbolName.Key()];
+    name = symbolName.Name();
+
+    // now we've resolved the name, let subscribed activities know
+    auto itSubscribedForSymbol = unresolvedTemplateInstantiationsPerSymbol_.find(symbolName.Key());
+    if (itSubscribedForSymbol != unresolvedTemplateInstantiationsPerSymbol_.end())
+    {
+        for (unsigned long long id : itSubscribedForSymbol->second)
+        {
+            auto itEntry = entries_.find(id);
+            
+            // may've been filtered out (didn't clean up the subscription as we're cleaning them all in a bit)
+            if (itEntry != entries_.end())
+            {
+                itEntry->second.Name = name;
+            }
+        }
+        itSubscribedForSymbol->second.clear();
+    }
+}
+
+void ExecutionHierarchy::OnCommandLine(const Activity& parent, const CommandLine& commandLine)
+{
+    auto it = entries_.find(parent.EventInstanceId());
+    assert(it != entries_.end());
+
+    it->second.Properties.try_emplace("Command Line", ToString(commandLine.Value()));
+}
+
+void ExecutionHierarchy::OnEnvironmentVariable(const Activity& parent, const EnvironmentVariable& environmentVariable)
+{
+    // we're not interested in all of them, only the ones that impact the build process
+    bool process = false;
+    if (parent.EventId() == EVENT_ID::EVENT_ID_COMPILER)
+    {
+        process = _wcsicmp(environmentVariable.Name(), L"CL") == 0
+               || _wcsicmp(environmentVariable.Name(), L"_CL_") == 0
+               || _wcsicmp(environmentVariable.Name(), L"INCLUDE") == 0
+               || _wcsicmp(environmentVariable.Name(), L"LIBPATH") == 0
+               || _wcsicmp(environmentVariable.Name(), L"PATH") == 0;
+    }
+    else if (parent.EventId() == EVENT_ID::EVENT_ID_LINKER)
+    {
+        process = _wcsicmp(environmentVariable.Name(), L"LINK") == 0
+               || _wcsicmp(environmentVariable.Name(), L"_LINK_") == 0
+               || _wcsicmp(environmentVariable.Name(), L"LIB") == 0
+               || _wcsicmp(environmentVariable.Name(), L"PATH") == 0
+               || _wcsicmp(environmentVariable.Name(), L"TMP") == 0;
+    }
+
+    if (process)
+    {
+        auto it = entries_.find(parent.EventInstanceId());
+        assert(it != entries_.end());
+        
+        it->second.Properties.try_emplace("Env Var: " + ToString(environmentVariable.Name()), ToString(environmentVariable.Value()));
+    }
+}
+
+void ExecutionHierarchy::OnFileInput(const Activity& parent, const FileInput& fileInput)
```

Comment:
Invocations can have many inputs and outputs. For example, a CL invocation can take many files and write many OBJs, and a linker invocation can take many input OBJs. Would it be possible to make the FileInput and FileOutput json properties an array instead of a single item?

Also, there is an inaccuracy in the documentation. Sometimes the parent of FileInput or FileOutput is a Pass1 activity, so I would replace `const Activity& parent` with `const Invocation& invocation`. in both `OnFileInput` and `OnFileOutput`. I'll make sure to change the documentation.

Reply:
We could make it an array instead of a single item, but the viewer will represent the `args` object as collapsible but would represent nested arrays and objects with its raw version.

This would be in the array form:
![example-grouped](https://user-images.githubusercontent.com/3693388/83572574-1fdce080-a52a-11ea-9e93-a35c80826acf.png)

While this would be if we separated them: 
![example-individual](https://user-images.githubusercontent.com/3693388/83572637-38e59180-a52a-11ea-8bad-d163d5209b2f.png)

What do you think?

We'll implement the parent `Invocation` part, no problem!

Reply:
In this case, using the separated approach is probably better. Thanks!